### PR TITLE
Protocol handler refactor and multithreading

### PR DESCRIPTION
**Pull request by dormando**
(will probably squash branch a bit prior to move).

This series is aimed at allowing mc-crusher to run in a more automated fashion, ie; a script manages the program and runs a list of tests. This required having timed runs, internal multithreading (vs me just running N of them), and an internal cleanup.

The protocol handlers are refactored to be much simpler and have a unified codepath for pipelining and preallocated buffers. Any command can now be pipelined up to N times, and new commands can be added much more easily.

### CODE REVIEW

File: mc-crusher.c [hunk 1]
```diff
@@ -180,305 +199,193 @@ static inline void run_counter(struct connection *c) {
...
-        vecs[i].iov_base = c->keys[*c->cur_key].key;
-        vecs[i].iov_len  = c->keys[*c->cur_key].key_len;
+        if (c->key_prealloc) {
+            vecs[i].iov_base = c->keys[*c->cur_key].key;
+            vecs[i].iov_len  = c->keys[*c->cur_key].key_len;
+        } else {
+            vecs[i].iov_base = c->wbuf_pos;
+            vecs[i].iov_len  = ascii_mget_format(c);
+        }
         run_counter(c);
```
Comment (BeyondSora) [inline mc-crusher.c]: since this function increments `cur_key` does it make sense to rename it accordingly? like `next_key`?

File: mc-crusher.c [hunk 2]
```diff
@@ -799,70 +680,121 @@ static void parse_config_line(char *line) {
         case PORT:
             template.port_num = atoi(value);
             break;
+        case THREAD:
```
Comment (dormando) [inline mc-crusher.c]: It bundles all of the connections for that line into its own thread. So if you say "conns=50,thread=true" you get 50 connections running that test in its own thread.

so, yep.

File: README [hunk 3]
```diff
@@ -34,6 +35,9 @@ Modify bench-sample to print what you are most interested in examining. Start
 multiple instances of mc-crusher if you think memcached can take more. In the
 future threads will be used for user convenience.
```
Comment (dormando) [inline README]: fixed

File: mc-crusher.c (hunk 1 above)
Comment (dormando) [inline mc-crusher.c]: maybe? it's always been that way. I may have initially thought it useful for other sorts of randomness in the tests, but it's only used for keys (ie; %'ing the counter to switch between command types)

**Comment by dormando** [comment]
I'll try to add more comments, sorry :) I left the individual commits in since that gives you smaller slices to look at to understand what I was changing. there're also comments in the commit messages.

1-4 are correct, except if you enable pipelining it piles in extra vectors to reduce syscalls.

there is also a separate reader for each test, which is only implemented as "blind read and throw on floor", but I'll be adding "check for END\r\n" and similar tests.

**Pull request merged**